java/ql/lib/semmle/code/java/frameworks/Strings.qll

```diff
@@ -11,19 +11,28 @@ private class StringSummaryCsv extends SummaryModelCsv {
         "java.lang;String;false;concat;(String);;Argument[0];ReturnValue;taint",
         "java.lang;String;false;concat;(String);;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;copyValueOf;;;Argument[0];ReturnValue;taint",
-        "java.lang;String;false;endsWith;;;Argument[-1];ReturnValue;taint",
+        "java.lang;String;false;describeConstable;;;Argument[-1];Element of ReturnValue;value",
         "java.lang;String;false;format;(Locale,String,Object[]);;Argument[1];ReturnValue;taint",
         "java.lang;String;false;format;(Locale,String,Object[]);;ArrayElement of Argument[2];ReturnValue;taint",
         "java.lang;String;false;format;(String,Object[]);;Argument[0];ReturnValue;taint",
         "java.lang;String;false;format;(String,Object[]);;ArrayElement of Argument[1];ReturnValue;taint",
         "java.lang;String;false;formatted;(Object[]);;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;formatted;(Object[]);;ArrayElement of Argument[0];ReturnValue;taint",
-        "java.lang;String;false;getChars;;;Argument[-1];Argument[2];taint",
         "java.lang;String;false;getBytes;;;Argument[-1];ReturnValue;taint",
+        "java.lang;String;false;getBytes;(int,int,byte[],int);;Argument[-1];Argument[2];taint",
+        "java.lang;String;false;getChars;;;Argument[-1];Argument[2];taint",
         "java.lang;String;false;indent;;;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;intern;;;Argument[-1];ReturnValue;taint",
-        "java.lang;String;false;join;;;Argument[0..1];ReturnValue;taint",
+        "java.lang;String;false;join;;;Argument[0..1];ReturnValue;taint", // TODO: ArrayElement of Argument?
```

Comment:
Should this here and maybe also in other cases use `ArrayElement of Argument` instead? What is the difference between tracking taint form an array (or varargs?) compared to tracking taint from the elements?

Reply:
Because `TaintTracking::Configuration` specifies an `allowImplicitRead` method that allows implicit reads from an array, collection or map-value (i.e., we can treat `y = propagateTaint(x)` or `sinkTaint(x)` as if they were `y = propagateTaint(x[n])` or `sinkTaint(x.get(n))`), in practice it doesn't make much difference for typical taint-tracking applications. In addition it specifies an `isAdditionalFlowStep` implementation that provides the opposite blurring of taint, saying that reading from a tainted map or tainted array yields tainted content.

In summary, for taint-tracking purposes `Field` and `SyntheticField` content continue to be faithfully distinguished, but array, collection and map content are blurred in both directions, with a tainted whole implying a tainted element and vice versa. You can specify them for clarity in the CSV description, but it won't make much (any?) difference to your results.